robot/src/main/java/us/ilite/robot/modules/Limelight.java

```diff
@@ -0,0 +1,250 @@
+        // h = hc - ht = Height of triangle for distance calculation
+        // d = distance from robot bumper to center of target bottom, this is what we're calculating
+        // db = measured distance from camera lens to robot bumper: SystemSettings.llCameraToBumperIn
+        // Ac = camera angle needed for calculating the distance: SystemSettings.llCameraAngleDeg
+        // ty = Vertical Offset From Crosshair To Target (-20.5 degrees to 20.5 degrees) parameter from the limelight
+
+        // we read the lime light values from mCurrentTarget, but this may be null if update is not
+        // called for the first time
+
+        double d = -1.0; // < 0 for error  TODO throw an execption???
+
+        if (this.mCurrentTarget != null) {
+            d = (SystemSettings.llCameraHeightIn - targetHeight) / 
+                Math.tan( SystemSettings.llCameraAngleDeg - this.mCurrentTarget.ty ) - 
+                SystemSettings.llCameraToBumperIn;
+        }
+
+        return d;
+    }
+
+    /**
+     * Calculate the distance to the currently tracked target by target type
+     * @param target
+     * @return Distance to target
+     */
+    public double calcTargetDistance( SystemSettings.VisionTarget target ) {
+        return this.calcTargetDistance( target.getHeight() );
+    }
+
+
+    /**
+     * Calculate the approach angle to the currently tracked target.
+     * A value of 0 deg means we are perpendicular to the target,
+     * A negative angle means the robot is to the left of the target
+     * A positive angle means the robot is to the right of the target
+     * @return Approach angle to target
+     */
+    public double calcTargetApproachAngle() {
+        // ts = ts angle or skew parameter from the limelight
+
+        // we read the lime light values from mCurrentTarget, but this may be null if update is not
+        // called for the first time
+
+        double approachAngle = 0.0;  
+        
+        // TODO throw an execption on error???
+        if (this.mCurrentTarget != null) {
+            // For -90 < Ts < -45 use the right hand function. For -45 < Ts <= 0 use the left hand function.
+
+
+            // get the skew angle and figure out which conversion to use
+            double ts = this.mCurrentTarget.ts;
+
+            if ( ts <= 0.0 && ts > -45.0 ) {
+                // left hand angle
+                approachAngle = - SystemSettings.llLeftACoeff + ts*SystemSettings.llLeftBCoeff + ts*Math.pow(SystemSettings.llLeftCCoeff, 2.0);
+            }
+            // TODO should we verify the  -90 < Ts < -45 for right hand angles and throw an exception if we don't meet it?
+            else { 
+                // right hand angle
+                approachAngle = SystemSettings.llRightACoeff + ts*SystemSettings.llRightBCoeff + ts*Math.pow(SystemSettings.llRightCCoeff, 2.0);
+            }
+
+        }
+
+        return approachAngle;
+    }
+
+    public class Point {
```

Comment:
This is fine for now.  We can refactor later to use the translate/rotate stuff from 254 once we start testing with the drivetrain.

Reply:
I'm using the Poof's Translation2D class.